.github/workflows/scripts/version_matrix.py

```
@@ -0,0 +1,51 @@
+    Reads project.requires-python, evaluates candidate versions against the
+    specifier, and returns the subset that satisfy the constraint, in ascending order.
+
+    Returns:
+        list[str]: The supported versions (e.g., ["3.11", "3.12"]).
+    """
+    with PYPROJECT.open("rb") as f:
+        data = tomllib.load(f)
+
+    spec_str = data.get("project", {}).get("requires-python")
+    if not spec_str:
+        raise KeyError("Missing 'project.requires-python' in pyproject.toml")
```

Ruff (0.14.7)

Avoid specifying long messages outside the exception class

(TRY003)